runtime/auth/aws-signing-standard/common/src/aws/smithy/kotlin/runtime/auth/awssigning/standard/Canonicalizer.kt

```diff
@@ -0,0 +1,215 @@
+ * @param requestString The canonical request string which is used in signature calculation
+ * @param signedHeaders A semicolon-delimited list of signed headers, all lowercase
+ * @param hash The hash for the body (either calculated from the body or pre-configured)
+ */
+internal data class CanonicalRequest(
+    val request: HttpRequestBuilder,
+    val requestString: String,
+    val signedHeaders: String,
+    val hash: String,
+)
+
+/**
+ * An object that can canonicalize a request.
+ */
+internal interface Canonicalizer {
+    companion object {
+        /**
+         * The default implementation of [Canonicalizer] to use.
+         */
+        val Default = DefaultCanonicalizer()
+    }
+
+    /**
+     * Canonicalize a request
+     * @param request The [HttpRequest] containing the data ready for signing
+     * @param config The signing parameters to use
+     * @param credentials Retrieved credentials used to canonicalize the request
+     */
+    suspend fun canonicalRequest(
+        request: HttpRequest,
+        config: AwsSigningConfig,
+        credentials: Credentials,
+    ): CanonicalRequest
+}
+
+// Taken from https://github.com/awslabs/aws-c-auth/blob/31d573c0dd328db5775f7a55650d27b8c08311ba/source/aws_signing.c#L118-L151
+private val skipHeaders = setOf(
+    "connection",
+    "sec-websocket-key",
+    "sec-websocket-protocol",
+    "sec-websocket-version",
+    "upgrade",
+    "user-agent",
+    "x-amzn-trace-id",
+)
+
+internal class DefaultCanonicalizer(private val sha256Supplier: HashSupplier = ::Sha256) : Canonicalizer {
+    private val logger = Logger.getLogger<DefaultCanonicalizer>()
+
+    override suspend fun canonicalRequest(
+        request: HttpRequest,
+        config: AwsSigningConfig,
+        credentials: Credentials,
+    ): CanonicalRequest {
+        val hash = when (config.bodyHash) {
+            is BodyHash.CalculateFromPayload -> request.body.calculateHash()
+            else -> config.bodyHash.hash!!
+        }
+
+        val signViaQueryParams = config.signatureType == AwsSignatureType.HTTP_REQUEST_VIA_QUERY_PARAMS
+        val addHashHeader = !signViaQueryParams && config.signedBodyHeader == AwsSignedBodyHeader.X_AMZ_CONTENT_SHA256
+        val sessionToken = credentials.sessionToken?.let { if (signViaQueryParams) it.urlEncodeComponent() else it }
+
+        val builder = request.toBuilder()
+
+        val params = when (config.signatureType) {
+            AwsSignatureType.HTTP_REQUEST_VIA_HEADERS -> builder.headers
+            AwsSignatureType.HTTP_REQUEST_VIA_QUERY_PARAMS -> builder.url.parameters
+            else -> TODO("Not yet implemented")
+        }
+        fun param(name: String, value: String?, predicate: Boolean = true) {
+            if (predicate && value != null) params[name] = value
+        }
+
+        param("Host", builder.url.host, !(signViaQueryParams || "Host" in params))
+        param("X-Amz-Algorithm", ALGORITHM_NAME, signViaQueryParams)
```

Comment:
It could probably be associated with `AwsSigningAlgorithm`. Not yet sure if SigV4a will have a single algorithm name or will use different ones depending on conditions. This code is likely to change once there's a second algorithm to support.